Autosummary additional links for inherited methods and attributes

satr-cowi opened the pull request:
Adds additional variables that can be used in autosummary templates for while also then splitting out methods and attributes if this is the preferred way of using.

### Feature or Bugfix
- Feature

### Purpose

The "path" is collected for the class from which this member is inherited, and then the name of the inherited method or attribute is appended to this depending on which it is. 

Can then be used in a template like so:
```
.. rubric:: {{ _('Inherited Methods') }}
.. autosummary::
	{% for item in inherited_methods %}
	{{item}}
	{%- endfor %}
```
Links then appear to where the method/attribute is actually defined.

Snapshot from the end result of my docs which use this looks like so:
![image](https://github.com/sphinx-doc/sphinx/assets/54358974/d03feb0a-ac7b-4c76-a382-7d434b845d47)

This is my first time contributing to a major opensource package, so do let me know if I've missed something or anything looks off.

satr-cowi committed:
Add links for inherited methods and attributes

satr-cowi committed:
Docs update for additional autosummary inherited options

Explanations of inherited_paths, inherited_methods and inherited_attributes

satr-cowi committed:
Merge branch 'sphinx-doc:master' into Autosummary_additional_links_for_inherited_methods_and_attributes

satr-cowi committed:
Formatting update for linter

picnixz commented on doc/usage/extensions/autosummary.rst:
```diff
@@ -303,6 +303,27 @@ The following variables are available in the templates:
+   List containing names of only inherited methods.  Only available for
+   classes.
+
+   .. versionadded:: 7.2.7
```
I think it'll be probably in 7.3 or 8.x (maybe you could leave a todo and we'll update when the PR is accepted).

satr-cowi replied:
Have put 7.3.0 for now; Will add a comment below to check this before merging

picnixz commented on doc/usage/extensions/autosummary.rst:
```diff
@@ -303,6 +303,27 @@ The following variables are available in the templates:
+
+   .. versionadded:: 7.2.7
+
+.. data:: inherited_paths
```
```suggestion
.. data:: inherited_qualname
```

We usually refer this path as the fully qualified name or sometimes the canonical name. We use `:canonical:` in Sphinx, so you could also have `canonical_inherited_names`.

satr-cowi replied:
Have adopeted qualnames

picnixz commented on doc/usage/extensions/autosummary.rst:
```diff
@@ -303,6 +303,27 @@ The following variables are available in the templates:
+
+.. data:: inherited_paths
+
+   List containing the full path inherited members. Will only return the closest
```
Maybe you should add an example to distinguish between this and the simple name.

satr-cowi replied:
Have added in a simple example to help with the various parameters

picnixz commented on sphinx/ext/autosummary/generate.py:
```diff
@@ -320,11 +320,34 @@ def generate_autosummary_content(name: str, obj: Any, parent: Any,
         ns['attributes'], ns['all_attributes'] = \
             _get_members(doc, app, obj, {'attribute', 'property'})
 
+        method_string = [m.split(".")[-1] for m in ns['methods']]
```
Perhaps it's better to have use `ns['all_methods']` since otherwise you'll only include public ones.

picnixz replied:
@satr-cowi Have you addressed this issue?

satr-cowi replied:
I think my preference is for ns['inherited_methods'] would be to only return the public methods (and therefore be consistent with ns['methods'])

We could introduce ns['all_inherited_methods'] as well, but the documentation doesn't include anything about ns['all_methods'] at the moment so we should probably include that too if we do this.

picnixz replied:
Personally, I want us to be consistent without breaking previous changes.

AFAIU the current behaviour is "ns[C] returns public members for category C" and "ns[all_C] returns all members for category C (including private ones)". I don't mind not documenting "all_*" for now, but I would like the documentation for ns["members"] & co. to be updated then. Currently it says (for "members"):

> List containing names of all members of the module or class. Only available
  for modules and classes.

In addition, your "inherited_qualnames" documentation tells me about: 

>    * ``members`` returns a list with many built in methods/attributes and
     ``['__init__', '_i_am_private', 'do_something', 'do_something_else',
     'foo_attr']``

So now your members also include private members.

satr-cowi replied:
Current behaviour (as in the example) is:

- methods and attributes returns just public methods/attributes
- members and inherited members return all members including private
- all_methods and all_attributes returns all methods/attributes including private, but these are not documented so most users won't know they exist

I've updated the "members" documentation to make it clear that this and "inherited members" do include private members. It was already clear that methods/attributes were only public.

The proposed additions are just that:

- inherited_methods and inherited_attributes returns just public inherited methods/attributes
- inherited_qualnames is the same as inherited_members (includes private), but just with the fully qualified names

Think this is pretty clear from the documentation now

picnixz commented on sphinx/ext/autosummary/generate.py:
```diff
@@ -320,11 +320,34 @@ def generate_autosummary_content(name: str, obj: Any, parent: Any,
+        ns['inherited_methods'] = []
+        ns['inherited_attributes'] = []
+
+        for i in (set(dir(obj)) - set(obj.__dict__.keys())):
```
Not all classes have a `__dict__` (and sometimes there are weird ones). So I think you should look at the implementation of `_get_members` to see how members are looked-up.

picnixz commented on sphinx/ext/autosummary/generate.py:
```diff
@@ -320,11 +320,34 @@ def generate_autosummary_content(name: str, obj: Any, parent: Any,
+            for cl in obj.__mro__:
+                local = set(cl.__dict__.keys())
+                if i in local:
+                    cl_str = cl.__module__ + "." + cl.__name__ + "." + i
```
Use f-strings instead.

satr-cowi replied:
Done

picnixz commented on sphinx/ext/autosummary/generate.py:
```diff
@@ -320,11 +320,34 @@ def generate_autosummary_content(name: str, obj: Any, parent: Any,
+                local = set(cl.__dict__.keys())
+                if i in local:
+                    cl_str = cl.__module__ + "." + cl.__name__ + "." + i
+                    if i[0] != "_":
```
I would use `startswith` instead. Also, what about special methods or those named `__foo__` for instance? because they would be excluded although they could be public.

satr-cowi replied:
Have removed this line, I now use the fact that "methods" and "attributes" are already filtered. "qualnames" now includes all the private/special methods too.

picnixz reviewed (commented):
- You'll need a CHANGES entry when the PR is ready.
- You'll need to add tests.

satr-cowi committed:
Review comment changes

Change name to qualnames
Re-use previously defined inherited members
Include private in qualnames, but still not in methods/attributes

satr-cowi committed:
Autosummary documentation improvements

satr-cowi committed:
Docs linter updates

satr-cowi committed:
Update autosummary.rst

satr-cowi committed:
Change inherited_members into ordered list

satr-cowi committed:
Added unittest to existing tests for inherited methods and attributes

satr-cowi committed:
Merge branch 'Autosummary_additional_links_for_inherited_methods_and_attributes' of https://github.com/satr-cowi/sphinx into Autosummary_additional_links_for_inherited_methods_and_attributes

satr-cowi committed:
lint formatting

satr-cowi committed:
Removed use of __dict__

Attributes now also include attributes defined in _init_ like in test

satr-cowi committed:
Test shouldn't depend on Python version or os. Includes attributes in __init__

satr-cowi committed:
docs updated

satr-cowi committed:
Linting corrections

satr-cowi committed:
correction

satr-cowi committed:
another linting correction

satr-cowi committed:
Refactor to minimise _get_class_members calls

satr-cowi committed:
Merge branch 'sphinx-doc:master' into Autosummary_additional_links_for_inherited_methods_and_attributes

satr-cowi committed:
Merge remote-tracking branch 'upstream/master' into Autosummary_additional_links_for_inherited_methods_and_attributes

satr-cowi committed:
Merge remote-tracking branch 'upstream/master' into Autosummary_additional_links_for_inherited_methods_and_attributes

satr-cowi committed:
Merge branch 'master' into Autosummary_additional_links_for_inherited_methods_and_attributes

satr-cowi committed:
Merge branch 'master' into Autosummary_additional_links_for_inherited_methods_and_attributes

satr-cowi committed:
Merge branch 'master' into Autosummary_additional_links_for_inherited_methods_and_attributes

picnixz commented on tests/test_extensions/test_ext_autosummary.py:
```diff
@@ -356,6 +356,49 @@ def test_autosummary_generate_content_for_module_imported_members_inherited_modu
+    import autosummary_dummy_inherited_module
+    template = Mock()
+
+    generate_autosummary_content('autosummary_dummy_inherited_module.InheritedAttrClass',
```
For this one, I would like to have a much wider test coverage. More precisely, I want to be able to test the behaviour of private names, mangled names, inherited methods that, static methods. For that, I would usggest creating a separate dummy module with multiple classes inside. I also want to see how multiple inheritance is handled and how shadowing names are done.

Most of the time, our tests cover simple cases which are fine but everytime there is an issue with autodoc, it's always because of some weird constructions. So I'd like to at least cover non-trivial cases (without trying really pathological cases).

satr-cowi replied:
Sure thing, I might need to lean a bit about these more niche cases but happy to give it a first go!

satr-cowi replied:
Have made a module where I have tried to test some of these features. Let me know if there is anything specific/additional that you think is missing.

picnixz commented on tests/test_extensions/test_ext_autosummary.py:
```diff
@@ -356,6 +356,49 @@ def test_autosummary_generate_content_for_module_imported_members_inherited_modu
+    assert context['module'] == 'autosummary_dummy_inherited_module'
+    assert context['objname'] == 'InheritedAttrClass'
+    assert context['name'] == 'InheritedAttrClass'
+    assert context['objtype'] == 'class'
```
Fully qualified names should also be tested (I don't see them but maybe I missed them?)

satr-cowi replied:
The full qualified name of this class is included as 'fullname', and the fully qualified name of the inherited members are in 'inherited_qualnames' (lines 383-389)

picnixz replied:
Oh my bad ! it didn't see it. Ideally I would prefer an exact match because that way you know whether we correctly picked up inherited members and how shadowed names were done, but I assume that there are too many things in the list.

satr-cowi replied:
I have tried including all the members in the new unit test, but you then get failures depending on the varying builtin methods included in different versions of Python (e.g. passes for 3.11 and 3.12, but fails on 3.10 and 3.13 for different reasons). Any thoughts on if there is a better way to deal with this rather than what I did above (i.e. don't worry about these additional items in the test)?

satr-cowi replied:
Have sorted using if statements depending on Python version as recommended. Implemented in a way where only small tweaks should be needed for future versions.

picnixz commented on doc/usage/extensions/autosummary.rst:
```diff
@@ -293,16 +293,79 @@ The following variables are available in the templates:
+          def do_something_else(self):
+              print("Bar something")
+
+   Some available parameters for the autosummary of class ``bar`` in the above
```
Also add an example for how to use it in a template.

satr-cowi replied:
Have added an example

satr-cowi committed:
Merge remote-tracking branch 'upstream/master' into Autosummary_additional_links_for_inherited_methods_and_attributes

satr-cowi committed:
ruff format

satr-cowi committed:
Only return the first instance of inheritance

satr-cowi committed:
Additional tests for more conditions

satr-cowi committed:
Clarification on members including private and template example

satr-cowi committed:
ruff format

satr-cowi committed:
lint corrections

satr-cowi committed:
Merge branch 'master' into Autosummary_additional_links_for_inherited_methods_and_attributes

satr-cowi committed:
Merge branch 'master' into Autosummary_additional_links_for_inherited_methods_and_attributes

picnixz commented on doc/usage/extensions/autosummary.rst:
```diff
@@ -293,16 +293,106 @@ The following variables are available in the templates:
 
    .. versionadded:: 1.8.0
 
+.. data:: inherited_qualnames
```
This should specify whether this includes public or private attributes.

satr-cowi replied:
Agreed, will clarify in the description that this does include private members (meaning methods and attributes)

picnixz commented on doc/usage/extensions/autosummary.rst:
```diff
@@ -293,16 +293,106 @@ The following variables are available in the templates:
 
-   List containing names of all members of the module or class.  Only available
-   for modules and classes.
+   List containing names of all members of the module or class including private
```
Just reconfirm, but this was already the case that it returned private attributes right? since we are doing `members = dir(self)`.

satr-cowi replied:
Yes, the existing attributes members and private members already included private ones. Therefore there shouldn't be any change here, just clarifying this in the docs.

picnixz reviewed (commented):
Ok, after having digested this PR I have one design concern. Currently `members` returns all attributes, whether they are private or not. In other words, I'd expect `inherited_XXX` to return all XXX whether they are private or not (because of the naming). I know that we already discussed the point with the methods not returning everything, but I honestly think we need to make the smallest breaking change as well as the simplest one and the most usable one.

For now, **do not change your implementation** (I'm just asking around).

In order to only have the public ones, I think it's better to have `public_inherited_XXX` and `inherited_XXX` where the former returns public lookalike attributes while the latter returns all XXX attributes, whether they are private or not.

What do you think? I'm sorry for my late review but I've been working more on CPython these past months, hence limiting my contribution to Sphinx itself.

---

cc @AA-Turner @chrisjsewell

satr-cowi committed:
add in version dependent attributes

satr-cowi committed:
ruff

satr-cowi committed:
update for 3_11 and remove extra term

satr-cowi committed:
add in members too

satr-cowi committed:
Simplify way to add new builtin members

satr-cowi committed:
tweak for 3_14 test

satr-cowi committed:
fix 3_14

satr-cowi committed:
correct order of members2

satr-cowi committed:
Improved comment

satr-cowi committed:
Improved docstring for inherited_qualnames

picnixz commented on sphinx/ext/autosummary/generate.py:
```diff
@@ -398,9 +427,12 @@ def _skip_member(app: Sphinx, obj: Any, name: str, objtype: str) -> bool:
 
 
-def _get_class_members(obj: Any) -> dict[str, Any]:
+def _get_class_members(obj: Any, return_object: bool = True) -> dict[str, Any]:
```
I would actually flip the boolean and use `return_types=False` by default. What do you think?

And whether we change the boolean or not, I think it would be good to make it keyword-only just so that we know what we do when we call the function (we don't need to go to its implementation).

satr-cowi replied:
No longer relevant, a different function is used instead as there is a new implementation of get class members

picnixz commented on tests/roots/test-ext-autosummary/autosummary_dummy_complex_inheritance_module.py:
```diff
@@ -0,0 +1,112 @@
+    def get_age(self):
+        return f"Parent class -  {self.age}"
+
+    # Test mangled name behaviour
```
Can we have a normal private attribute/method (no mangled, just simple underscore)?

satr-cowi replied:
Have added

picnixz commented on tests/roots/test-ext-autosummary/autosummary_dummy_complex_inheritance_module.py:
```diff
@@ -0,0 +1,112 @@
+
+# The following is used to process the expected builtin members for different
+# versions of Python. The base list is for v3.11 (the latest testing when added)
+# and new builtin attributes/methods in later versions can be appended below.
```
Can we have those values declared in the test instead? the test roots should be only contains the code we "test" (so this file should be like a real one on which you run autodoc).

satr-cowi replied:
Moved to test function

picnixz commented on tests/test_extensions/test_ext_autosummary.py:
```diff
@@ -488,6 +488,376 @@ def test_autosummary_generate_content_for_module_imported_members_inherited_modu
+        '__setattr__',
+        '__sizeof__',
+        '__str__',
+        '__subclasshook__',
```
Aren't those methods inherited from `object`? can't they be automatically detected like this?

satr-cowi replied:
Couldn't use object directly as it didn't have some of the expected attributes/methods, but a using a empty class did the trick.

picnixz commented on tests/test_extensions/test_ext_autosummary.py:
```diff
@@ -488,6 +488,376 @@ def test_autosummary_generate_content_for_module_imported_members_inherited_modu
+        'builtins.object.__sizeof__',
+        'builtins.object.__str__',
+        'builtins.object.__subclasshook__',
+    ]
```
Can't we generate the `builtins.object.*` methods using `[f'builtins.object.{name}' for name in dir(object)]`?

satr-cowi replied:
Similarly, have simplified this into a generated list with some tweaks where necessary

satr-cowi committed:
Merge remote-tracking branch 'upstream/master' into Autosummary_additional_links_for_inherited_methods_and_attributes

satr-cowi committed:
Merge remote-tracking branch 'upstream/master' into Autosummary_additional_links_for_inherited_methods_and_attributes

satr-cowi committed:
Incorporate changes from latest version

satr-cowi committed:
correct_typo

satr-cowi committed:
ruff format

satr-cowi committed:
try adding to index

satr-cowi committed:
switch order

satr-cowi committed:
update tests for new call

satr-cowi committed:
try this for mypy

satr-cowi committed:
use an empty base class to test against

satr-cowi committed:
autogenerate attr too

satr-cowi committed:
annotation built in

satr-cowi committed:
specific version lists

satr-cowi committed:
addition of normal private methods too

satr-cowi committed:
Merge branch 'master' into Autosummary_additional_links_for_inherited_methods_and_attributes

satr-cowi committed:
Merge branch 'sphinx-doc:master' into Autosummary_additional_links_for_inherited_methods_and_attributes
